### ISSUE TITLE
Check point a process inside a container

### ISSUE THREAD

**Issue by washraf** [xemul/criu#103]
I am using this to checkpoint a process inside a docker container: 
      docker exec source criu dump -t 8568 --images-dir /criu --shell-job

##The Result is:

Warn  (ptrace.c:81): Unable to interrupt task: 8568 (Operation not permitted)
Error (ptrace.c:53): Unable to detach from 8568: No such process
Error (cr-dump.c:1578): Dumping FAILED.

I am sure the process is correct and its id is too.

**Comment by xemul** [comment]
You start criu in restricted environment -- kernel doesn't allow even to ptrace-attach to the process. Can you check this by trying to strace the process in the same manner?

**Comment by xemul** [comment]
OK, re-open the issue if it still stands.

**Comment by koliaskostas** [comment]
I am also interested on this issue.

I have a similar problem where I am running apache from within a container and try to checkpoint as root.

**Comment by xemul** [comment]
@koliaskostas , the error you have is

    Warn (criu/ptrace.c:82): Unable to interrupt task: 1 (Operation not permitted)

this is EPERM reported from ptrace(PTRACE_SEIZE...) which is purely about kernel restrictions. If you cannot strace the task you want to dump, then criu will not be able to work too.

**Comment by koliaskostas** [comment]
OK I had some progress with this issue:

It seems that in docker processes run in unprivileged mode so if you forcefully (and wrongly in most of the cases) want to change that simply use the --privileged flag in the docker run command.
More info here: https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities

This will eliminate the error  `Unable to interrupt task` but still won't solve the problem of checkpointing.

Keep also in mind that strace will still not be able to run, giving an error
```
strace: attach: ptrace(PTRACE_ATTACH, ...): Operation not permitted
Could not attach to process.  If your uid matches the uid of the target
process, check the setting of /proc/sys/kernel/yama/ptrace_scope, or try
again as the root user.  For more details, see /etc/sysctl.d/10-ptrace.conf
```
To deal with that you will probably need to set flags like --cap-add SYS_PTRACE

Finally, based on some research two  flags are good to be used in such cases (and only for testing purposes) --security-opt seccomp:unconfined  --security-opt apparmor:unconfined as the two have been found to cause problems.

so basically I ended up running:
`docker run -t -i -d --cap-add SYS_PTRACE --security-opt seccomp:unconfined  --security-opt apparmor:unconfined --privileged mine/ubuntu-apache4 /bin/bash`

and then 
`docker exec -t -i 777ec4e0b232655a4b65ad5cb3bd0a2a6edde6a92ce45135c753557b1349bc7f /bin/bash`

started apache service and attempted to checkpoint it without a problem.

Now restoring it...that's another issue.

**Comment by xemul** [comment]
@koliaskostas , your set of options for Docker container looks like a ... pretty big hammer. The resulting container has much more than required to just strace the process... But, anyway, you're moving forward, which is good.
